File: polyglot/python/signature_database_management.py
import os
import json
from collections import defaultdict

class SignatureDatabase:
    def __init__(self, db_path='signature_db.json'):
        self.db_path = db_path
        self.signatures = self._load_database()

    def _load_database(self):
        if not os.path.exists(self.db_path):
            return defaultdict(list)
        try:
            with open(self.db_path, 'r') as f:
                return json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            return defaultdict(list)

    def _save_database(self):
        with open(self.db_path, 'w') as f:
            json.dump(self.signatures, f, indent=2)

    def add_signature(self, file_type, signature):
        if file_type not in self.signatures:
            self.signatures[file_type] = []
        self.signatures[file_type].append(signature)
        self._save_database()

    def remove_signature(self, file_type, signature):
        if file_type in self.signatures and signature in self.signatures[file_type]:
            self.signatures[file_type].remove(signature)
            self._save_database()

    def list_signatures(self):
        return self.signatures

    def clear_database(self):
        self.signatures = defaultdict(list)
        self._save_database()


def main():
    db = SignatureDatabase()
    print("Current Signatures:", db.list_signatures())

    # Example: Add some signatures
    db.add_signature("JPEG", "FFD8FF")
    db.add_signature("PNG", "89504E47")
    db.add_signature("ZIP", "504B0304")
    print("After Adding Signatures:", db.list_signatures())

    # Example: Remove a signature
    db.remove_signature("JPEG", "FFD8FF")
    print("After Removing JPEG Signature:", db.list_signatures())

    # Example: Clear database
    db.clear_database()
    print("After Clearing Database:", db.list_signatures())


if __name__ == "__main__":
    main()

File: polyglot/ruby/signature_database_management.rb
# polyglot/ruby/signature_database_management.rb

require 'set'

module Filecarve
  module SignatureDatabase
    class Database
      attr_reader :signatures, :file_types

      def initialize
        @signatures = {}
        @file_types = Set.new
      end

      def add_signature(signature_name, pattern, file_type)
        raise ArgumentError, "Signature name cannot be empty" if signature_name.empty?
        raise ArgumentError, "Pattern cannot be empty" if pattern.empty?

        unless pattern.is_a?(String) && pattern.start_with?(/\A(?:(?:\d+|\w+|\W)+)\z/)
          raise ArgumentError, "Invalid pattern format"
        end

        @signatures[signature_name] = {
          pattern: pattern,
          file_type: file_type
        }
        @file_types.add(file_type)
      end

      def remove_signature(signature_name)
        if @signatures.key?(signature_name)
          @file_types.delete(@signatures[signature_name][:file_type])
          @signatures.delete(signature_name)
        else
          raise KeyError, "Signature '#{signature_name}' not found"
        end
      end

      def list_signatures
        @signatures.keys
      end

      def list_file_types
        @file_types.to_a
      end

      def find_matching_signatures(data)
        matches = {}
        @signatures.each do |name, sig|
          if data.include?(sig[:pattern])
            matches[name] = sig[:file_type]
          end
        end
        matches
      end
    end

    # Example usage
    def self.run_demo
      db = Database.new

      puts "Adding signatures..."
      db.add_signature("txt", "text", "Text File")
      db.add_signature("jpg", "jpeg", "JPEG Image")
      db.add_signature("png", "png", "PNG Image")

      puts "\nList of signatures:"
      puts db.list_signatures.join(', ')

      puts "\nList of file types:"
      puts db.list_file_types.join(', ')

      puts "\nTesting signature matching..."
      test_data = "This is a sample text file with some text content."
      matches = db.find_matching_signatures(test_data)
      puts "Matches found: #{matches.empty? ? 'none' : matches.size}"

      puts "\nRemoving 'txt' signature..."
      db.remove_signature("txt")

      puts "\nList of signatures after removal:"
      puts db.list_signatures.join(', ')
    end
  end
end

# Run the demo if this file is executed directly
if __FILE__ == $0
  Filecarve::SignatureDatabase.run_demo
end